builder/amazon/common/access_config.go

```diff
@@ -237,6 +305,177 @@ func (c *AccessConfig) IsChinaCloud() bool {
+			creds, err = c.GetCredentialsFromSession()
+			if err != nil {
+				return nil, err
+			}
+		} else {
+			return nil, fmt.Errorf("Error loading credentials for AWS Provider: %w", err)
+		}
+	} else {
+		log.Printf("[INFO] AWS Auth provider used: %q", cp.ProviderName)
+	}
+
+	// In the "normal" flow (i.e. not assuming a role), we return here.
+	if c.AssumeRole.AssumeRoleARN == "" {
+		return creds, nil
+	}
+
+	// create a config for the assume role session based off the config we
+	// created for our main sessions
+	assumeRoleAWSConfig := config.Copy()
+	assumeRoleAWSConfig.CredentialsChainVerboseErrors = aws.Bool(true)
+
+	assumeRoleSession, err := session.NewSession(assumeRoleAWSConfig)
+
+	if err != nil {
+		return nil, fmt.Errorf("error creating assume role session: %w", err)
+	}
+
+	stsclient := sts.New(assumeRoleSession)
+	assumeRoleProvider := &stscreds.AssumeRoleProvider{
+		Client:  stsclient,
+		RoleARN: c.AssumeRole.AssumeRoleARN,
+	}
+
+	if c.AssumeRole.AssumeRoleDurationSeconds > 0 {
+		assumeRoleProvider.Duration = time.Duration(c.AssumeRole.AssumeRoleDurationSeconds) * time.Second
+	}
+
+	if c.AssumeRole.AssumeRoleExternalID != "" {
+		assumeRoleProvider.ExternalID = aws.String(c.AssumeRole.AssumeRoleExternalID)
+	}
+
+	if c.AssumeRole.AssumeRolePolicy != "" {
+		assumeRoleProvider.Policy = aws.String(c.AssumeRole.AssumeRolePolicy)
+	}
+
+	if len(c.AssumeRole.AssumeRolePolicyARNs) > 0 {
+		var policyDescriptorTypes []*sts.PolicyDescriptorType
+
+		for _, policyARN := range c.AssumeRole.AssumeRolePolicyARNs {
+			policyDescriptorType := &sts.PolicyDescriptorType{
+				Arn: aws.String(policyARN),
+			}
+			policyDescriptorTypes = append(policyDescriptorTypes, policyDescriptorType)
+		}
+
+		assumeRoleProvider.PolicyArns = policyDescriptorTypes
+	}
+
+	if c.AssumeRole.AssumeRoleSessionName != "" {
+		assumeRoleProvider.RoleSessionName = c.AssumeRole.AssumeRoleSessionName
+	}
+
+	if len(c.AssumeRole.AssumeRoleTags) > 0 {
+		var tags []*sts.Tag
+
+		for k, v := range c.AssumeRole.AssumeRoleTags {
+			tag := &sts.Tag{
+				Key:   aws.String(k),
+				Value: aws.String(v),
+			}
+			tags = append(tags, tag)
+		}
+
+		assumeRoleProvider.Tags = tags
+	}
+
+	if len(c.AssumeRole.AssumeRoleTransitiveTagKeys) > 0 {
+		assumeRoleProvider.TransitiveTagKeys = aws.StringSlice(c.AssumeRole.AssumeRoleTransitiveTagKeys)
+	}
+
+	providers = []awsCredentials.Provider{assumeRoleProvider}
+
+	assumeRoleCreds := awsCredentials.NewChainCredentials(providers)
+
+	_, err = assumeRoleCreds.Get()
+	if err != nil {
+		return nil, fmt.Errorf("Unable to assume role: %w", err)
+	}
+
+	return assumeRoleCreds, nil
+}
+
+// GetCredentialsFromSession returns credentials derived from a session. A
+// session uses the AWS SDK Go chain of providers so may use a provider (e.g.,
+// ProcessProvider) that is not part of the Terraform provider chain.
+func (c *AccessConfig) GetCredentialsFromSession() (*awsCredentials.Credentials, error) {
```

Comment:
If Packer's custom build credentials chain doesn't return valid credentials, This function gives a fallback to the default credentials chain that Packer used to rely on. This is necessary if the credentials get inferred from instance metadata if it's running inside AWS.